dGame/dCinema/Recorder.cpp

```diff
@@ -0,0 +1,1195 @@
+	}
+
+	recorder->AddRecord(new PlayEffectRecord(std::to_string(effectId)));
+}
+
+MovementRecord::MovementRecord(const NiPoint3& position, const NiQuaternion& rotation, const NiPoint3& velocity, const NiPoint3& angularVelocity, bool onGround, bool dirtyVelocity, bool dirtyAngularVelocity) {
+	this->position = position;
+	this->rotation = rotation;
+	this->velocity = velocity;
+	this->angularVelocity = angularVelocity;
+	this->onGround = onGround;
+	this->dirtyVelocity = dirtyVelocity;
+	this->dirtyAngularVelocity = dirtyAngularVelocity;
+}
+
+void MovementRecord::Act(Entity* actor) {
+	auto* controllableComponent = actor->GetComponent<ControllablePhysicsComponent>();
+
+	if (controllableComponent) {
+		controllableComponent->SetPosition(position);
+		controllableComponent->SetRotation(rotation);
+		controllableComponent->SetVelocity(velocity);
+		controllableComponent->SetAngularVelocity(angularVelocity);
+		controllableComponent->SetIsOnGround(onGround);
+		controllableComponent->SetDirtyVelocity(dirtyVelocity);
+		controllableComponent->SetDirtyAngularVelocity(dirtyAngularVelocity);
+	}
+
+	Game::entityManager->SerializeEntity(actor);
+}
+
+void MovementRecord::Serialize(tinyxml2::XMLDocument& document, tinyxml2::XMLElement* parent) {
+	auto* element = document.NewElement("MovementRecord");
+
+	element->SetAttribute("x", position.x);
+	element->SetAttribute("y", position.y);
+	element->SetAttribute("z", position.z);
+
+	element->SetAttribute("qx", rotation.x);
+	element->SetAttribute("qy", rotation.y);
+	element->SetAttribute("qz", rotation.z);
+	element->SetAttribute("qw", rotation.w);
+
+	element->SetAttribute("vx", velocity.x);
+	element->SetAttribute("vy", velocity.y);
+	element->SetAttribute("vz", velocity.z);
+
+	element->SetAttribute("avx", angularVelocity.x);
+	element->SetAttribute("avy", angularVelocity.y);
+	element->SetAttribute("avz", angularVelocity.z);
+
+	element->SetAttribute("g", onGround);
+
+	element->SetAttribute("dv", dirtyVelocity);
+
+	element->SetAttribute("dav", dirtyAngularVelocity);
+
+	element->SetAttribute("t", m_Delay);
+
+	parent->InsertEndChild(element);
+}
+
+void MovementRecord::Deserialize(tinyxml2::XMLElement* element) {
+	position.x = element->FloatAttribute("x");
+	position.y = element->FloatAttribute("y");
+	position.z = element->FloatAttribute("z");
+
+	rotation.x = element->FloatAttribute("qx");
+	rotation.y = element->FloatAttribute("qy");
+	rotation.z = element->FloatAttribute("qz");
+	rotation.w = element->FloatAttribute("qw");
+
+	velocity.x = element->FloatAttribute("vx");
+	velocity.y = element->FloatAttribute("vy");
+	velocity.z = element->FloatAttribute("vz");
+
+	angularVelocity.x = element->FloatAttribute("avx");
+	angularVelocity.y = element->FloatAttribute("avy");
+	angularVelocity.z = element->FloatAttribute("avz");
+
+	onGround = element->BoolAttribute("g");
+
+	dirtyVelocity = element->BoolAttribute("dv");
+
+	dirtyAngularVelocity = element->BoolAttribute("dav");
```

Comment:
Why are we saving dirty variables here?

Reply:
In my experience, handling a positional packet and not passing on the dirty flags can have some odd effects. I see that's still how we handle it. Don't think the redundency hurts.

Reply:
Well, setting the values as they are above already sets these flags to dirty in the contexts where they exist, hence seeming pointless to bother saving this because you should always mark hard changes to velocity as dirty